added new remove mode (like in android notification area)

imbryk opened the pull request:
Hi,

I added new way to remove items - it mimics the way in which notifications are removed from the notification area.
The item can be flinged or moved aside to be removed, and it can be done on the whole item, not just a drag handle.

While removing the item you can't move it up or down, and vice versa, when you're moving the item vertically, you cannot remove it. Because of that, this new mode works only with mDragInitMode set ON_DRAG (actually if you set this remove mode, the mDragInitMode is ignored and behaves as if it was set to ON_DRAG).

Additionaly the TestBedDSLV activity in demo app is modified, so it starts with this new remove mode turned on.

imbryk committed:
added FLING_OR_SLIDE_REMOVE mode

swancheng commented:
Thank you very much! 

------------------ 原始邮件 ------------------

发件人: "Leszek Mzyk"notifications@github.com;
发送时间: 2013年1月11日(星期五) 晚上10:51
收件人: "bauerca/drag-sort-listview"drag-sort-listview@noreply.github.com; 

主题: [drag-sort-listview] added new remove mode (like in androidnotification area) (#61)

Hi,
I added new way to remove items - it mimics the way in which notifications are removed from the notification area.The item can be flinged or moved aside to be removed, and it can be done on the whole item, not just a drag handle.
While removing the item you can't move it up or down, and vice versa, when you're moving the item vertically, you cannot remove it. Because of that, this new mode works only with mDragInitMode set ON_DRAG (actually if you set this remove mode, the mDragInitMode is ignored and behaves as if it was set to ON_DRAG).
You can merge this Pull Request by running  git pull https://github.com/imbryk/drag-sort-listview fling-or-slide-remove
Or view, comment on, or merge it at:
  https://github.com/bauerca/drag-sort-listview/pull/61
Commit Summary
added FLING_OR_SLIDE_REMOVE modeFile Changes
M demo/res/values/strings.xml (1)
M demo/src/com/mobeta/android/demodslv/TestBedDSLV.java (6)
M library/src/com/mobeta/android/dslv/DragSortController.java (117)
M library/src/com/mobeta/android/dslv/DragSortListView.java (91)Patch Links:
https://github.com/bauerca/drag-sort-listview/pull/61.patch
https://github.com/bauerca/drag-sort-listview/pull/61.diff

imbryk committed:
added styleable flingOrSlide

bauerca commented:
I just merged your additions locally and played around with the demo. Some thoughts:

flingOrSlide could work with everything. For example:
1. flingOrSlide remove and onDown init sort: If a down event is detected on the drag handle, a drag sort starts. Otherwise, if a side swipe gesture is detected that originated on the non-drag-handle part of the item, just fling/slide remove is enabled.
2. flingOrSlide remove and onDrag init sort: Done! :)
3. flingOrSlide remove and onLongPress init sort: Trivial. Long press on drag handle starts drag-sort, swipe anywhere on item does remove thing.

For all of the above, a user could also flingOrSlide to remove while drag-sorting. The following would help with this:

A new dslv attr called `dslv:remove_during_sort_enabled="true | false"` (defaulting to `true`). If `false`, item removals can only happen if the user shows no intention to sort (e.g. does only a side swipe on a single item). In this case, the floating View is locked in y (can't sort). If the user _does_ show intention to sort (by moving finger up or down after drag initiation), then the floating View is locked in x and removes can't happen. If `true`, the only difference from `false` is that during a sort, removals are enabled; i.e. the floating View is not locked in x.

I just noticed that click remove is broken. The top item is always removed. I also got the following exception after click removing twice in a row (this is in the DSLVTestBed):

```
E/AndroidRuntime( 6001): java.lang.ArrayIndexOutOfBoundsException
E/AndroidRuntime( 6001):    at java.util.ArrayList.get(ArrayList.java:313)
E/AndroidRuntime( 6001):    at android.widget.ArrayAdapter.getItem(ArrayAdapter.java:298)
E/AndroidRuntime( 6001):    at com.mobeta.android.demodslv.DSLVFragment$2.remove(DSLVFragment.java:41)
```

Lots to do...

imbryk commented:
1 and 3 - your'e right, down or longPress on handle will initiate sorting, while the flingOrSlide removal will happen only on the rest of the item. I didn't think of it, because at the beginning both sorting and removing was only available on handle - I added the possibility to remove on whole item later.

As for removing while sorting - I find it annoying and wanted to remove it (that's why I needed to check which way user moves finger), but probably someone might think different, so it's true that putting the flag will solve the problem.

I will make the changes on monday and we'll see.

As for the error - I must check what happened - thanks for pointing.

imbryk committed:
fixed error in  onClick remove mode

imbryk committed:
flingOrSlide working with onDown and onLongPress drag init modes

imbryk commented:
I'm sorry I have not written anything for a whole week, but I had to stay home with a sick child.

If you could check this modifications - I didn't add the `remove_during_sort_enabled` yet, but I think that if it is added, then it should work the same way for all the remove modes. I will try to make it this way.

edit - there are still a few bugs ;)

bauerca commented:
Okay thanks for the update, I'll check it out. You should read #64. If there are no complaints to it, I would like to take your work as a starting point for the consolidation. Thanks.

imbryk commented:
great - so I'll stop working on `remove_during_sort_enabled`, and clean the current code a bit.

I've got one more proposal - right now the removal happens on the whole item, while the sorting happens on `drag_handle_id` view. I would add another attr `fling_handle_id`, if the handle id is not set, then the removal will happen on the whole item (like it is happening now), otherwise it will happen only on the handle. The drag handle will have priority to fling handle (e.g. if it is child of the fling handle or if they overlap).

imbryk committed:
added attr fling_handle_id, cleaned code a bit

imbryk committed:
remove methods consolidation

imbryk commented:
In the newest commit I left only click and fling remove modes - you're welcome to work with it (or maybe I could help with something more?)

Leszek

bauerca commented:
I won't have time to look at this for a while. Busy with other work. Looking fwd to it, though! Thanks.

bauerca commented:
This is some great stuff. I can't find any bugs.

A few minor requests. I've been trying to adhere to the android coding style. I notice that your tabs are two spaces, whereas the rest of dslv is 4-space tabs. Could you fix this? If you want, you can also update the README to reflect your changes, but I'm happy to do this. I will definitely write the latest news item and acknowledgement at the top.

Probably not a big deal, but I noticed that a fling remove won't work if it originates on the drag handle and the drag init mode is onlongpress. It could.

Thanks again! I'll happily pull when at least the tabs issues is resolved.

imbryk committed:
Applied android formatting. Fling remove works on the drag handle when drag init mode is onLongPress

imbryk commented:
ok - I've corrected the formating (using android-formatting.xml), and added fling remove on handle while drag init mode is onlongpress.

As for README - maybe you should update it, as english is your native language and you'll write it better ;)

Leszek

bauerca commented:
That's great, thanks. I'm happy to do the readme.

bauerca commented:
Finally. Was in the midst of a move to Portland.
